CMakeLists.txt

```diff
@@ -31,33 +31,46 @@ elseif (WIN32)
 endif ()
 
 # Will be overwritten during the cmake initialization.
+
+# Value can be one of "None" and "ControlFlow"
+# If None, no LVI mitigation will be applied
+# If ControlFlow, only control-flow mitigations will be applied
+# Support of All LVI hardening (-mlvi-hardening) is omitted
 set(LVI_MITIGATION
     "None"
-    CACHE STRING "Build with LVI mitigation.")
+    CACHE STRING "Build with LVI mitigation. Options: [None, ControlFlow]")
 set(LVI_MITIGATION_BINDIR
     "None"
-    CACHE STRING "Path to the LVI mitigation bindir.")
+    CACHE STRING "[clang-10 only] Path to the LVI mitigation bindir.")
 
 # On Unix, configuring cmake to use the preferred compiler.
 # Note that this must be done before the `project` command.
 if (UNIX)
-  if (LVI_MITIGATION MATCHES ControlFlow)
-    # If the LVI mitigation is enabled, use the customized compilation toolchain.
+  # For the normal build, if the CC environment variable has been specified or if
+  # the CMAKE_C_COMPILER cmake variable has been passed to cmake, use the C compiler
+  # that has been specified. Otherwise, prefer clang. Same for C++ compiler.
+  if (NOT DEFINED ENV{CC} AND NOT DEFINED CMAKE_C_COMPILER)
+    find_program(CMAKE_C_COMPILER clang-11 clang-10 clang)
+  endif ()
+  if (NOT DEFINED ENV{CXX} AND NOT DEFINED CMAKE_CXX_COMPILER)
+    find_program(CMAKE_CXX_COMPILER clang++-11 clang++-10 clang++)
+  endif ()
+
+  if (CMAKE_C_COMPILER MATCHES "10$" OR CMAKE_CXX_COMPILER MATCHES "10$")
+    set(OE_USING_CLANG_10 TRUE)
+  else ()
+    set(OE_USING_CLANG_10 FALSE)
+  endif ()
+
+  # For clang-10, check options for Load Value Injection (LVI) mitigation
+  if (OE_USING_CLANG_10 AND LVI_MITIGATION MATCHES ControlFlow)
     if (${LVI_MITIGATION_BINDIR} MATCHES None)
       message(FATAL_ERROR "LVI_MITIGATION_BINDIR is not specified.")
     endif ()
+
+    # If the LVI mitigation is enabled, use the customized compilation toolchain.
```

Comment:
Can you double check if `configure_lvi_mitigation_build` overwrite `CMAKE_C_COMPILER` and `CMAKE_CXX_COMPILER` as expected?

Reply:
I tested with main branch on OpenEnclave, it doesn't seem to correctly override the compiler used. Even if this line of code is called: https://github.com/openenclave/openenclave/blob/master/cmake/configure_lvi_mitigation_build.cmake#L69

Reply:
It might be the reason that we use if and else to separate the two cases.

Reply:
```
SET CMAKE_C_COMPILER to ./lvi_mitigation_bin/clang-10
CMAKE_C_COMPILER: /usr/bin/clang-10
```
Even with `if else` its still reverts back to the default compiler...

Reply:
Can you see if that works on the master branch?

Reply:
> I tested with main branch on OpenEnclave, it doesn't seem to correctly override the compiler used. Even if this line of code is called: https://github.com/openenclave/openenclave/blob/master/cmake/configure_lvi_mitigation_build.cmake#L69

Doesn't seem so on my machine.

Reply:
```
$ cmake .. -DLVI_MITIGATION=ControlFlow -DLVI_MITIGATION_BINDIR=/home/mingwei/lvi_mitigation_bin
-- The C compiler identification is Clang 10.0.1
-- The CXX compiler identification is Clang 10.0.1
-- The ASM compiler identification is Clang
-- Found assembler: /home/mingwei/lvi_mitigation_bin/clang-10
-- Check for working C compiler: /home/mingwei/lvi_mitigation_bin/clang-10
-- Check for working C compiler: /home/mingwei/lvi_mitigation_bin/clang-10 -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Check for working CXX compiler: /home/mingwei/lvi_mitigation_bin/clang++-10
-- Check for working CXX compiler: /home/mingwei/lvi_mitigation_bin/clang++-10 -- works
```

Reply:
This was merged into the Clang-11 working branch, and I made some modifications to address the comments here. Diff: https://github.com/openenclave/openenclave/commit/f36abd814639fd9393c9b5bd250713c510d2d120#diff-1e7de1ae2d059d21e1dd75d5812d5a34b0222cef273b7c3a2af62eb747f9d20a

@mingweishih could you confirm the below is the expected behaviour with what you proposed? The output was obtained from running the changes in CI:

Clang 10, ControlFlow (ControlFlow-GNU)
```
+ cmake /home/oeadmin/workspace/Private/OpenEnclave/Azure-Linux -G Ninja -DCMAKE_INSTALL_PREFIX:PATH=/opt/openenclave -DCPACK_GENERATOR=DEB -DCODE_COVERAGE=OFF -DUSE_DEBUG_MALLOC=ON -DCMAKE_BUILD_TYPE=RelWithDebInfo -DLVI_MITIGATION_BINDIR=/usr/local/lvi-mitigation/bin -DLVI_MITIGATION=ControlFlow -DLVI_MITIGATION_SKIP_TESTS=OFF -Wdev
-- The C compiler identification is Clang 10.0.0
-- The CXX compiler identification is Clang 10.0.0
-- The ASM compiler identification is Clang
-- Found assembler: /usr/local/lvi-mitigation/bin/clang-10
-- Check for working C compiler: /usr/local/lvi-mitigation/bin/clang-10
-- Check for working C compiler: /usr/local/lvi-mitigation/bin/clang-10 -- works
```
Clang 10, ControlFlow-Clang (expected failure)
```
+ cmake /home/oeadmin/workspace/Private/OpenEnclave/Azure-Linux -G Ninja -DCMAKE_INSTALL_PREFIX:PATH=/opt/openenclave -DCPACK_GENERATOR=DEB -DCODE_COVERAGE=OFF -DUSE_DEBUG_MALLOC=ON -DCMAKE_BUILD_TYPE=RelWithDebInfo -DLVI_MITIGATION_BINDIR=/usr/local/lvi-mitigation/bin -DLVI_MITIGATION=ControlFlow-Clang -DLVI_MITIGATION_SKIP_TESTS=OFF -Wdev
-- The C compiler identification is Clang 10.0.0
-- The CXX compiler identification is Clang 10.0.0
-- The ASM compiler identification is Clang
-- Found assembler: /usr/bin/clang-10
-- Check for working C compiler: /usr/bin/clang-10
-- Check for working C compiler: /usr/bin/clang-10 -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Check for working CXX compiler: /usr/bin/clang++-10
-- Check for working CXX compiler: /usr/bin/clang++-10 -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
CMake Error at CMakeLists.txt:95 (message):
  ControlFlow-Clang only works with clang version >= 11.  But got
  /usr/bin/clang-10
```
Clang-11, ControlFlow
```
+ cmake /home/oeadmin/workspace/Private/OpenEnclave/Azure-Linux -G Ninja -DCMAKE_INSTALL_PREFIX:PATH=/opt/openenclave -DCPACK_GENERATOR=DEB -DCODE_COVERAGE=OFF -DUSE_DEBUG_MALLOC=ON -DCMAKE_BUILD_TYPE=Debug -DLVI_MITIGATION_BINDIR=/usr/local/lvi-mitigation/bin -DLVI_MITIGATION=ControlFlow -DLVI_MITIGATION_SKIP_TESTS=ON -Wdev
-- The C compiler identification is Clang 11.1.0
-- The CXX compiler identification is Clang 11.1.0
-- The ASM compiler identification is Clang
-- Found assembler: /usr/local/lvi-mitigation/bin/clang-11
-- Check for working C compiler: /usr/local/lvi-mitigation/bin/clang-11
-- Check for working C compiler: /usr/local/lvi-mitigation/bin/clang-11 -- works
-- Detecting C compiler ABI info
```
Clang-11, ControlFlow-Clang
```
+ cmake /home/oeadmin/workspace/Private/OpenEnclave/Azure-Linux -G Ninja -DCMAKE_INSTALL_PREFIX:PATH=/opt/openenclave -DCPACK_GENERATOR=DEB -DCODE_COVERAGE=OFF -DUSE_DEBUG_MALLOC=ON -DCMAKE_BUILD_TYPE=RelWithDebInfo -DLVI_MITIGATION_BINDIR=/usr/local/lvi-mitigation/bin -DLVI_MITIGATION=ControlFlow-Clang -DLVI_MITIGATION_SKIP_TESTS=OFF -Wdev
-- The C compiler identification is Clang 11.1.0
-- The CXX compiler identification is Clang 11.1.0
-- The ASM compiler identification is Clang
-- Found assembler: /usr/bin/clang-11
-- Check for working C compiler: /usr/bin/clang-11
-- Check for working C compiler: /usr/bin/clang-11 -- works
```

Reply:
The behaviors are as what I expected. Thanks for doing this!